opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,285 @@
+    """A module for opsdroid to allow memory to persist in matrix room state."""
+
+    def __init__(self, config, opsdroid=None):
+        """Start the database connection."""
+        super().__init__(config, opsdroid=opsdroid)
+        self.name = "matrix"
+        self.room = config.get("default_room", "main")
+        self._single_state_key = config.get("single_state_key", True)
+        self.should_encrypt = config.get("should_encrypt", True)
+        self._event_type = "dev.opsdroid.database"
+        self.should_migrate = True
+
+        _LOGGER.debug("Loaded matrix database connector.")
+
+    @property
+    def connector(self):
+        """Define the initiated matrix connector."""
+        return self.opsdroid.get_connector("matrix")
+
+    @property
+    def room_id(self):
+        """Define ID of the room to store data in."""
+        return self.connector.lookup_target(self.room)
+
+    @property
+    def is_room_encrypted(self):
+        """Check if room is encrypted."""
+        return (
+            self.connector._allow_encryption
+            and self.room_id in self.connector.connection.store.load_encrypted_rooms()
+        )
+
+    async def migrate_database(self):
+        """Migrate existing 'opsdroid.database' state event to 'dev.opsdroid.database' event."""
+
+        data = await self.connector.connection.room_get_state(room_id=self.room_id)
+        if isinstance(data, RoomGetStateError):
+            _LOGGER.error(
+                f"Error migrating from opsdroid.database to {self._event_type} in room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+
+        if not any(event["type"] == self._event_type for event in data.events):
+            for event in data.events:
+                if event["type"] == "opsdroid.database" and event["content"]:
+                    await self.connector.connection.room_put_state(
+                        room_id=self.room_id,
+                        event_type=self._event_type,
+                        content=event["content"],
+                        state_key=event["state_key"],
+                    )
+                    await self.connector.connection.room_redact(
+                        room_id=self.room_id, event_id=event["event_id"],
+                    )
+
+        self.should_migrate = False
+
+    async def _verify_and_get(self, state_key, value_dict):
+        """Get state and verify duplicate entries."""
+
+        ori_data = await self.connector.connection.room_get_state_event(
+            room_id=self.room_id, event_type=self._event_type, state_key=state_key,
+        )
+        if isinstance(ori_data, RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error getting the state event of type {self._event_type} with "
+                f"state_key '{state_key}' from matrix room {self.room_id}: "
+                f"{ori_data.message}({ori_data.status_code})"
+            )
+            return
```

Comment:
Given my [comment](https://github.com/opsdroid/opsdroid/pull/1587/files#r474656706) below, shouldn't this be 
```suggestion
            return {}
```

Reply:
Also, this isn't always an error, if this is being triggered by the first PUT in a room, meaning logging this at error is probably a bad idea under all situations.

Reply:
We won't get a `RoomGetStateEventError` for the first put in the room, that would happen for other problems like connection errors or membership errors etc